File: data_structure/set/src/SetDemo.go
// Go
// データ構造: セット (Set)

package main

import (
	"fmt"
)

type SetData struct {
	data []interface{}
}

func NewSetData() *SetData {
	return &SetData{
		data: []interface{}{},
	}
}

func (s *SetData) Get() []interface{} {
	// 要素を取得
	return s.data
}

func (s *SetData) GetIndex(item interface{}) int {
	// 指定された要素がセット内に存在するかどうかをチェックする。
	for i, v := range s.data {
		if v == item {
			return i
		}
	}
	fmt.Printf("ERROR: %v は範囲外です\n", item)
	return -1
}

func (s *SetData) GetValue(index int) interface{} {
	// 指定されたインデックスの要素を取得する。
	if 0 <= index && index < len(s.data) {
		return s.data[index]
	} else {
		fmt.Printf("ERROR: %v は範囲外です\n", index)
		return nil
	}
}

func (s *SetData) Add(item interface{}) bool {
	// 要素をセットに追加する。
	for _, v := range s.data {
		if v == item {
			fmt.Printf("ERROR: %v は重複です\n", item)
			return false
		}
	}
	s.data = append(s.data, item)
	return true
}

func (s *SetData) Remove(item interface{}) bool {
	// 指定された要素をセットから削除する。
	for i, v := range s.data {
		if v == item {
			// スライスから要素を削除
			s.data = append(s.data[:i], s.data[i+1:]...)
			return true
		}
	}
	fmt.Printf("ERROR: %v は範囲外です\n", item)
	return false
}

func (s *SetData) IsEmpty() bool {
	// 空かどうかをチェックする
	return len(s.data) == 0
}

func (s *SetData) Size() int {
	// 要素数を返す
	return len(s.data)
}

func (s *SetData) Clear() bool {
	// 空にする
	s.data = []interface{}{}
	return true
}

func main() {
	fmt.Println("Set TEST -----> start")

	fmt.Println("\nnew")
	setData := NewSetData()
	fmt.Printf("  現在のデータ: %v\n", setData.Get())

	fmt.Println("\nadd")
	input := []interface{}{10, 20, 30, 20, 40}
	for _, item := range input {
		fmt.Printf("  入力値: %v\n", item)
		output := setData.Add(item)
		fmt.Printf("  出力値: %v\n", output)
		fmt.Printf("  現在のデータ: %v\n", setData.Get())
	}

	fmt.Println("\nsize")
	output := setData.Size()
	fmt.Printf("  出力値: %v\n", output)

	fmt.Println("\nis_empty")
	outputEmpty := setData.IsEmpty()
	fmt.Printf("  出力値: %v\n", outputEmpty)

	fmt.Println("\nget_value")
	inputIndexes := []int{0, 2, 5}
	for _, index := range inputIndexes {
		fmt.Printf("  入力値: %v\n", index)
		outputValue := setData.GetValue(index)
		fmt.Printf("  出力値: %v\n", outputValue)
	}

	fmt.Println("\nget_index")
	inputItems := []interface{}{30, 99}
	for _, item := range inputItems {
		fmt.Printf("  入力値: %v\n", item)
		outputIndex := setData.GetIndex(item)
		fmt.Printf("  出力値: %v\n", outputIndex)
	}

	fmt.Println("\nremove")
	removeItems := []interface{}{20, 50, 10}
	for _, item := range removeItems {
		fmt.Printf("  入力値: %v\n", item)
		outputRemove := setData.Remove(item)
		fmt.Printf("  出力値: %v\n", outputRemove)
		fmt.Printf("  現在のデータ: %v\n", setData.Get())
	}

	fmt.Println("\nsize")
	outputSize := setData.Size()
	fmt.Printf("  出力値: %v\n", outputSize)

	fmt.Println("\nclear")
	outputClear := setData.Clear()
	fmt.Printf("  出力値: %v\n", outputClear)
	fmt.Printf("  現在のデータ: %v\n", setData.Get())

	fmt.Println("\nis_empty")
	outputEmpty = setData.IsEmpty()
	fmt.Printf("  出力値: %v\n", outputEmpty)

	fmt.Println("\nSet TEST <----- end")
}

File: data_structure/set/src/SetDemo.rs
// Rust
// データ構造: セット (Set)

use std::collections::HashSet;

struct SetData {
    data: HashSet<i32>,
}

impl SetData {
    fn new() -> Self {
        SetData {
            data: HashSet::new(),
        }
    }

    fn get(&self) -> Vec<i32> {
        self.data.iter().cloned().collect()
    }

    fn get_index(&self, item: &i32) -> Option<usize> {
        self.data.iter().position(|&x| x == *item)
    }

    fn get_value(&self, index: usize) -> Option<i32> {
        self.get().get(index).cloned()
    }

    fn add(&mut self, item: i32) -> bool {
        self.data.insert(item)
    }
    
    fn remove(&mut self, item: &i32) -> bool {
        self.data.remove(item)
    }
    
    fn is_empty(&self) -> bool {
        self.data.is_empty()
    }
    
    fn size(&self) -> usize {
        self.data.len()
    }
    
    fn clear(&mut self) {
        self.data.clear()
    }
}

fn main() {
    println!("Set TEST -----> start");

    println!("\nnew");
    let mut set_data = SetData::new();
    println!("  現在のデータ: {:?}", set_data.get());

    println!("\nadd");
    let input = [10, 20, 30, 20, 40];
    for &item in &input {
        println!("  入力値: {}", item);
        let output = set_data.add(item);
        println!("  出力値: {}", output);
        println!("  現在のデータ: {:?}", set_data.get());
    }

    println!("\nsize");
    let output = set_data.size();
    println!("  出力値: {}", output);

    println!("\nis_empty");
    let output = set_data.is_empty();
    println!("  出力値: {}", output);

    println!("\nget_value");
    let input = [0, 2, 5];
    for &index in &input {
        println!("  入力値: {}", index);
        let output = set_data.get_value(index);
        println!("  出力値: {:?}", output);
    }

    println!("\nget_index");
    let input = [30, 99];
    for &item in &input {
        println!("  入力値: {}", item);
        let output = set_data.get_index(&item);
        println!("  出力値: {:?}", output);
    }

    println!("\nremove");
    let input = [20, 50, 10];
    for &item in &input {
        println!("  入力値: {}", item);
        let output = set_data.remove(&item);
        println!("  出力値: {}", output);
        println!("  現在のデータ: {:?}", set_data.get());
    }
    
    println!("\nsize");
    let output = set_data.size();
    println!("  出力値: {}", output);

    println!("\nclear");
    set_data.clear();
    println!("  現在のデータ: {:?}", set_data.get());

    println!("\nis_empty");
    let output = set_data.is_empty();
    println!("  出力値: {}", output);

    println!("\nSet TEST <----- end");
}
